Store delegators in pool state

teodanciu opened the pull request:
# Description

Store, in each pool's StakePoolState, the stake credentials that have delegated to that pool. 
This set is updated in DELEG rule, for both shelley and conway: 

- on (re)delegation
- when the stake credential is unregistered 

The set is used to improve the performance of removing delegations when retiring the pool. 
Based on PR: https://github.com/IntersectMBO/cardano-ledger/pull/5330

Closes: https://github.com/IntersectMBO/cardano-ledger/issues/4990


<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [ ] Commits in meaningful sequence and with useful messages.
- [ ] Tests added or updated when needed.
- [ ] `CHANGELOG.md` files updated for packages with externally visible changes.  
      **NOTE: _New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd)).
- [ ] Versions updated in `.cabal` and `CHANGELOG.md` files when necessary, according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] Version bounds in `.cabal` files updated when necessary.  
      **NOTE: _If bounds change in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process)).
- [x] Code formatted (use `scripts/fourmolize.sh`).
- [x] Cabal files formatted (use `scripts/cabal-format.sh`).
- [x] CDDL files are up to date (use `scripts/gen-cddl.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) updated (use `scripts/gen-hie.sh`).
- [ ] Self-reviewed the diff.

teodanciu committed:
Update `StakePoolState` delegs when (re)delegating stake in conway

teodanciu committed:
Refactor `processDRepUndelgation`

for consistency

teodanciu committed:
Update checks in `DelegSpec` to include delegs in `StakePoolState`

and add a test case for redelegating drep

teodanciu committed:
Change state of shelley `DelegRule` to `CertState`

teodanciu committed:
Update `StakePoolState` delegs in shelley

teodanciu committed:
Update shelley `Transition` to set the pool delegations

and add a test for it

teodanciu committed:
Update conway `Transition` to set the pool delegations

and add a test for it

teodanciu committed:
Intern stake credentials in `StakePoolState` CBOR decoder

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/State/StakePool.hs:
```diff
@@ -132,6 +137,8 @@ data StakePoolState = StakePoolState
   -- ^ Optional metadata (URL and hash)
   , spsDeposit :: !(CompactForm Coin)
   -- ^ Deposit for each pool
+  , spsDelegs :: !(Set (Credential 'Staking))
```
I don't think we are saving much by shortening, but by expanding the name we are reducing ambiguity.
```suggestion
  , spsDelegators :: !(Set (Credential 'Staking))
```
I wish the same thought came to me when I was adding such field for DReps. Once this PR is merged, I'll create a ticket to rename the field for DReps as well for consistency

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/State/StakePool.hs:
```diff
@@ -162,6 +169,9 @@ spsMetadataL = lens spsMetadata $ \sps md -> sps {spsMetadata = md}
 spsDepositL = lens spsDeposit $ \sps d -> sps {spsDeposit = d}
 
+spsDelegsL :: Lens' StakePoolState (Set (Credential 'Staking))
+spsDelegsL = lens spsDelegs $ \sps delegs -> sps {spsDelegs = delegs}
```
Ditto
```suggestion
spsDelegatorsL :: Lens' StakePoolState (Set (Credential 'Staking))
spsDelegatorsL = lens spsDelegators $ \sps delegs -> sps {spsDelegators = delegs}
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/State/Account.hs:
```diff
@@ -164,7 +163,7 @@ isAccountRegistered cred accounts = Map.member cred (accounts ^. accountsMapL)
   EraAccounts era =>
   (AccountState era -> AccountState era) -> Credential 'Staking -> Accounts era -> Accounts era
-adjustAccountState cred f = accountsMapL %~ Map.adjust cred f
+adjustAccountState f cred = accountsMapL %~ Map.adjust f cred
```
Nice catch :smile:

lehins commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/Examples/Combinators.hs:
```diff
@@ -369,7 +378,11 @@ reapPool pool cs = cs {chainNes = nes'}
-    ds' = ds {dsAccounts = removeStakePoolDelegations (Set.singleton poolId) accounts'}
+    delegsToClear =
+      foldMap spsDelegs $
+        maybeToList $
```
You don't really need this step, Maybe also has a Foldable instance :wink: 
```suggestion
```

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -1637,20 +1638,51 @@ expectDelegatedToPool ::
+      expectNothingExpr (accountState ^. stakePoolDelegationAccountStateL)
+    assertBool
+      ("Expected no stake pool state delegation to contain the stake credential: " <> show cred)
+      (all (Set.notMember cred . spsDelegs) (Map.elems pools))
```
`Map` too has a `Foldable` instance :smiley: 
```suggestion
      (all (Set.notMember cred . spsDelegs) pools)
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/PoolReap.hs:
```diff
@@ -226,6 +227,11 @@ poolReapTransition = do
 
+    delegsToClear cState pools =
+      foldMap spsDelegs $
+        Map.elems $
```
Ditto
```suggestion
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Transition.hs:
```diff
@@ -409,14 +410,20 @@ shelleyRegisterInitialAccounts ::
+    registerAndDelegate ((stakeKeyHash, stakePool), ptr) (!accounts, !stakePoolMap)
       | stakePool `Map.member` stakePools =
-          registerShelleyAccount (KeyHashObj stakeKeyHash) ptr deposit (Just stakePool) accounts
+          ( (registerShelleyAccount (KeyHashObj stakeKeyHash) ptr deposit (Just stakePool) accounts)
```
Redundant parenthesis
```suggestion
          ( registerShelleyAccount (KeyHashObj stakeKeyHash) ptr deposit (Just stakePool) accounts
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Pool.hs:
```diff
@@ -293,9 +294,14 @@ poolDelegationTransition = do
+          let futureStakePoolState =
+                mkStakePoolState
+                  (stakePoolState ^. spsDepositL)
+                  (stakePoolState ^. spsDelegsL)
```
This is actually incorrect, because we need to use the value of the delegators Set not from when certificate was submitted, but when the actual rotation of future to current happens. This is because by the time future becomes current new delegations might have been added and old ones removed! In other words we need to update this in the POOLREAP instead: https://github.com/IntersectMBO/cardano-ledger/blob/f776625c3ddf1bd0a3873d1683265b97e6733953/eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/PoolReap.hs#L155
```haskell
psStakePools = 
  Map.unionWith
   (\newPoolState oldPoolState -> newPoolState {spsDelegators = spsDelegators oldPoolState)
   (ps0 ^. psFutureStakePoolsL)
   (ps0 ^. psStakePoolsL)
```

```suggestion
                  mempty
```
It would be very good to add a test case that exercises this scenario.

In the matter of fact, in order to avoid mistakes like this in the future, I think we need to switch back to `PoolParams`:
```diff
-  , psFutureStakePools :: !(Map (KeyHash 'StakePool) StakePoolState)
+  , psFutureStakePools :: !(Map (KeyHash 'StakePool) PoolParams)
```
https://github.com/IntersectMBO/cardano-ledger/blob/f776625c3ddf1bd0a3873d1683265b97e6733953/libs/cardano-ledger-core/src/Cardano/Ledger/State/CertState.hs#L229

I've created a ticket about this: #5343

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs:
```diff
@@ -397,5 +418,21 @@ updateReservesAndTreasury targetPot combinedMap available ds = do
+  where
+    stakePoolToUnDeleg oldPool
+      | Just oldPool /= mNewPool = Just oldPool
+      | otherwise = Nothing
```
To simplify a bit and promote safety (having two arguments with the same type is always a bit dangerous, since accidental swapping of arguments is a thing)
We also get the newStakePool passed as well in the arguments, why not update it in the same swoop
With all these changes, I believe it will be nice to put this function in the same module where `PState` is defined, instead of here, since can also be reused in Conway
```suggestion
-- Update reverse StakePool delegation Set whenever stake credential unregisters or changes its delegation. Whenever new delegation matches the old one this is a noop
unDelegReDelegStakePool ::
  Credential 'Staking ->
  AccountState era ->
  -- ^ Account that is loosing its current delegation and/or acquiring a new one
  Maybe (KeyHash 'StakePool) ->
  -- ^ Potential new delegation. In case when stake credential unregisters this must be `Nothing`.
  PState era ->
  PState era
unDelegReDelegStakePool stakeCred accountState mNewStakePool = 
  fromMaybe (psStakePoolsL %~ addNewDelegation) $ do
    curStakePool <- accountState ^. stakePoolDelegationAccountStateL
    pure $ 
      -- There is no need to update set of delegations if delegation is unchanged
      if Just curStakePool == mNewStakePool
      then id
      else 
        psStakePoolsL %~ addNewDelegation . Map.adjust (spsDelegsL %~ Set.delete stakeCred) curStakePool)
  where
    addNewDelegation = 
      maybe id (Map.adjust (spsDelegsL %~ Set.insert stakeCred)) mNewStakePool
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs:
```diff
@@ -242,34 +247,57 @@ instance
+        Nothing -> do
+          failBecause $ StakeDelegationImpossibleDELEG cred
+          pure certState
+        Just currentAccount ->
```
There is no other type of account state except current, so I think it makes sense to drop the prefix
```suggestion
        Just accountState ->
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs:
```diff
@@ -242,34 +247,57 @@ instance
+              & certPStateL %~ \ps ->
+                ps
+                  & unDelegStakePool cred (currentAccount ^. stakePoolDelegationAccountStateL) (Just stakePool)
+                  & psStakePoolsL %~ Map.adjust (spsDelegsL %~ Set.insert cred) stakePool
```
With my other suggestions
```suggestion
              & certPStateL %~ unDelegReDelegStakePool cred accountState (Just stakePool)
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs:
```diff
@@ -242,34 +247,57 @@ instance
-      pure (ds & accountsL %~ registerShelleyAccount cred ptr compactDeposit Nothing)
+      pure $ certState & certDStateL . accountsL %~ registerShelleyAccount cred ptr compactDeposit Nothing
     UnRegTxCert cred -> do
       let !(mAccountState, !accounts) = unregisterShelleyAccount cred (ds ^. accountsL)
```
Might as well make it strict, since with new changes it has to be forced regardless
```suggestion
      let !(!mAccountState, !accounts) = unregisterShelleyAccount cred (ds ^. accountsL)
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs:
```diff
@@ -242,34 +247,57 @@ instance
+            certState
+              & certDStateL . accountsL .~ accounts
+              & certPStateL
+                %~ unDelegStakePool cred (accountState ^. stakePoolDelegationAccountStateL) Nothing
```
```suggestion
                %~ unDelegReDelegStakePool cred accountState Nothing
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -255,8 +256,10 @@ conwayDelegTransition = do
       isJust mAccountState ?! StakeKeyNotRegisteredDELEG stakeCred
       failOnJust checkStakeKeyHasZeroRewardBalance StakeKeyHasNonZeroRewardAccountBalanceDELEG
       pure $
```
I suggest we do here the same thing we did in Shelley:
```suggestion
      failOnJust checkStakeKeyHasZeroRewardBalance StakeKeyHasNonZeroRewardAccountBalanceDELEG
      case mAccountState of
        Nothing -> do
          failBecause $ StakeKeyNotRegisteredDELEG stakeCred
          pure certState
        Just accountState ->
          pure $ 
            certState
              & certDStateL . accountsL .~ newAccounts
              & certVStateL %~ unDelegDRep stakeCred mCurDelegatee
              & certPStateL %~ unDelegReDelegStakePool cred accountState Nothing
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -255,8 +256,10 @@ conwayDelegTransition = do
+        certState
+          & certDStateL . accountsL .~ newAccounts
+          & certVStateL %~ unDelegDRep stakeCred mCurDelegatee
+          & certPStateL %~ unDelegStakePool stakeCred mCurDelegatee Nothing
```
See suggestion above
```suggestion
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -312,11 +315,13 @@ processDelegationInternal preserveIncorrectDelegation stakeCred mCurDelegatee ne
       cState
         & certDStateL . accountsL
           %~ adjustAccountState (stakePoolDelegationAccountStateL ?~ stakePool) stakeCred
+        & certPStateL %~ adjustPState stakePool
```
This will require `processDelegationInternal` to accept `Maybe (AccountState era)`, instead of `Maybe Delegatee`, which should be totally doable
```suggestion
        & certPStateL %~ unDelegReDelegStakePool cred accountState (Just stakePool)
```

teodanciu replied:
Modified to account for the fact that the accountState that we have is optional, yet  in `unDelegReDelegStakePool` it is not optional.

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -326,23 +331,39 @@ processDelegationInternal preserveIncorrectDelegation stakeCred mCurDelegatee ne
+      DelegStakeVote oldPool _ | Just oldPool /= mNewPool -> Just oldPool
+      _ -> Nothing
+
+unDelegDRep ::
```
Instead of `unDelegDRep` I propose to add the exact same function as I suggested for stake pools in the Shelley DELEG rule, but for DReps and move it to where VState is defined:

```haskell
-- Update reverse DRep delegation Set whenever stake credential unregisters or changes its delegation. Whenever new delegation matches the old one this is a noop
unDelegReDelegDRep ::
  Credential 'Staking ->
  AccountState era ->
  -- ^ Account that is loosing its current delegation and/or acquiring a new one
  Maybe DRep ->
  -- ^ Potential new delegation. In case when stake credential unregisters this must be `Nothing`.
  VState era ->
  VState era
unDelegReDelegDRep stakeCred accountState mNewDRep = 
  fromMaybe (vsDRepsL %~ addNewDelegation) $ do
    dRep@(DRepCredential dRepCred) <- accountState ^. dRepDelegationAccountStateL
    pure $ 
      -- There is no need to update set of delegations if delegation is unchanged
      if Just dRep == mNewStakePool
      then id
      else 
        vsDRepsL %~ addNewDelegation . Map.adjust (drepDelegsL %~ Set.delete stakeCred) dRepCred)
  where
    addNewDelegation = 
      case mNewDRep of
        Just (DRepCredential dRepCred) ->
          Map.adjust (drepDelegsL %~ Set.insert stakeCred) dRepCred
        Nothing -> id
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -326,23 +331,39 @@ processDelegationInternal preserveIncorrectDelegation stakeCred mCurDelegatee ne
             _ -> cState'
+    adjustPState newPool =
+      (psStakePoolsL %~ Map.adjust (spsDelegsL %~ Set.insert stakeCred) newPool)
+        . unDelegStakePool stakeCred mCurDelegatee (Just newPool)
```
No longer needed
```suggestion
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -312,11 +315,13 @@ processDelegationInternal preserveIncorrectDelegation stakeCred mCurDelegatee ne
+            cState
               & certDStateL . accountsL
                 %~ adjustAccountState (dRepDelegationAccountStateL ?~ dRep) stakeCred
+              & certVStateL %~ unDelegDRep stakeCred mCurDelegatee
```
```suggestion
              & certVStateL %~ unDelegReDelDRep stakeCred mAccountState (Just dRep)
```

teodanciu replied:
I modified this a little, in order to have  `unDelegReDelDRep` take an `AccountState`, rather than a `Maybe AccountState` -   in order to keep the semantics of the function: deal with the existing dreps when reversing the delegation.

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -312,11 +315,13 @@ processDelegationInternal preserveIncorrectDelegation stakeCred mCurDelegatee ne
+              & certVStateL %~ unDelegDRep stakeCred mCurDelegatee
           dReps
             | preserveIncorrectDelegation = cState ^. certVStateL . vsDRepsL
             | otherwise = cState' ^. certVStateL . vsDRepsL
```
This will need to be slightly rewritten:
```suggestion
      in if preserveIncorrectDelegation
         then cState & certVStateL %~ case dRep of 
                DRepCredential dRepCred ->
                  vsDRepsL %~ Map.adjust (drepDelegsL %~ Set.insert stakeCred) dRepCred
                _ -> id
         else cState'
```
And all the stuff below is no longer needed.

teodanciu replied:
I modified this suggestion, because accountState has to be modifed also in the `preserveIncorrectDelegation` state

lehins reviewed (changes requested):
This is a great piece of work. There is only one real problem in this PR and the rest of the comments are things that I've notice that can be improved thanks to your new changes.

teodanciu committed:
Refactor PState update logic to handle stake undelegation/redelegation

uniformly across eras

teodanciu committed:
Refactor VState update logic to handle dRep undelegation/redelegation

teodanciu committed:
Set delegators of updated pools in `PoolReap` rule

in order not to lose the delegations that happened after the update,
before the end of the epoch

teodanciu commented:
I addressed your suggestions, @lehins  - whenever you have time to have another look.

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs:
```diff
@@ -327,46 +351,39 @@ delegationTransition = do
-                      }
+                    certState
+                      & certDStateL . dsIRewardsL . iRDeltaReservesL %~ (<> invert (toDeltaCoin coin))
+                      & certDStateL . dsIRewardsL . iRDeltaTreasuryL %~ (<> toDeltaCoin coin)
```
There is a special operator for this in microlens :slightly_smiling_face: 
```suggestion
                      & certDStateL . dsIRewardsL . iRDeltaReservesL <>~ invert (toDeltaCoin coin)
                      & certDStateL . dsIRewardsL . iRDeltaTreasuryL <>~ toDeltaCoin coin
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs:
```diff
@@ -327,46 +351,39 @@ delegationTransition = do
-                      }
+                    certState
+                      & certDStateL . dsIRewardsL . iRDeltaReservesL %~ (<> toDeltaCoin coin)
+                      & certDStateL . dsIRewardsL . iRDeltaTreasuryL %~ (<> invert (toDeltaCoin coin))
```
Ditto
```suggestion
                      & certDStateL . dsIRewardsL . iRDeltaReservesL <>~ toDeltaCoin coin
                      & certDStateL . dsIRewardsL . iRDeltaTreasuryL <>~ invert (toDeltaCoin coin)
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/DelegSpec.hs:
```diff
@@ -282,7 +304,13 @@ spec = do
-        ifBootstrap (expectNotDelegatedVote cred) (expectDelegatedVote cred (DRepCredential drepCred2))
+        ifBootstrap
+          ( do
+              -- we cannot `expectNotDelegatedVote` because the the delegation is still in the DRepState of the other pool
```
```suggestion
              -- we cannot `expectNotDelegatedVote` because the delegation is still in the DRepState of the other pool
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/DelegSpec.hs:
```diff
@@ -282,7 +304,13 @@ spec = do
+              -- we cannot `expectNotDelegatedVote` because the the delegation is still in the DRepState of the other pool
+              accounts <- getsNES $ nesEsL . esLStateL . lsCertStateL . certDStateL . accountsL
+              expectNothingExpr (lookupDRepDelegation cred accounts)
+          )
```
Just to make sure we fully preserve the bug:
```suggestion
              dReps <- getsNES $ nesEsL . esLStateL . lsCertStateL . certVStateL . vsDRepsL
              drepState2 <- expectJust $ Map.lookup drepCred2 dReps
              drepDelegs drepState2 `shouldSatisfy` Set.member cred
          )
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -252,16 +253,23 @@ conwayDelegTransition = do
-          certState & certDStateL . accountsL .~ newAccounts
+      case mAccountState of
+        Nothing ->
+          do
```
```suggestion
        Nothing -> do
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -252,16 +253,23 @@ conwayDelegTransition = do
+              & certPStateL %~ unDelegReDelegStakePool stakeCred accountState Nothing
     ConwayDelegCert stakeCred delegatee -> do
-      mCurDelegatee <- checkStakeKeyIsRegistered stakeCred
+      mAccountState <- checkStakeKeyIsRegistered stakeCred
```
You know, this binding is only used once. Let's just inline in for clarity:
```suggestion
      let mAccountState = lookupAccountState stakeCred accounts
      isJust mAccountState ?! StakeKeyNotRegisteredDELEG stakeCred
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs:
```diff
@@ -312,37 +319,26 @@ processDelegationInternal preserveIncorrectDelegation stakeCred mCurDelegatee ne
+              cState
+                & certDStateL . accountsL %~ adjustAccountState (dRepDelegationAccountStateL ?~ dRep) stakeCred
+                & certVStateL %~ insertDRepDeleg dRep
+            else cState'
```
Let's get rid of some duplication and that ugly tick:
```suggestion
       in cState
            & certDStateL . accountsL %~ adjustAccountState (dRepDelegationAccountStateL ?~ dRep) stakeCred
            & maybe
                (certVStateL %~ insertDRepDeleg dRep)
                (\accountState -> certVStateL %~ unDelegReDelegDRep stakeCred accountState (Just dRep))
                (guard (not preserveIncorrectDelegation) >> mAccountState)
```
or this way maybe a bit clearer :shrug: 
```suggestion
       in cState
            & certDStateL . accountsL %~ adjustAccountState (dRepDelegationAccountStateL ?~ dRep) stakeCred
            & case mAccountState of
                Just accountState
                  | not preserveIncorrectDelegation -> 
                  certVStateL %~ unDelegReDelegDRep stakeCred accountState (Just dRep)
                Nothing ->
                  certVStateL %~ insertDRepDeleg dRep
```

teodanciu replied:
Nice, thanks.

lehins reviewed (approved):
This looks great!! Thank you for your work!

teodanciu committed:
Add delegators to StakePoolState

teodanciu committed:
Use StakePoolState delegs to remove delegations when retiring pools
